storage/repository/generic.go

```diff
@@ -0,0 +1,218 @@
+package repository
+
+import (
+	"io"
+	"strings"
+
+	"fyne.io/fyne"
+)
+
+// splitNonEmpty works exactly like strings.Split(), but only returns non-empty
+// components.
+func splitNonEmpty(str, sep string) []string {
+	components := []string{}
+	for _, v := range strings.Split(str, sep) {
+		if len(v) > 0 {
+			components = append(components, v)
+		}
+	}
+	return components
+}
+
+// GenericParent can be used as a common-case implementation of
+// HierarchicalRepository.Parent(). It will create a parent URI based on
+// IETF RFC3986.
+//
+// In short, the URI is separated into it's component parts, the path component
+// is split along instances of '/', and the trailing element is removed. The
+// result is concatenated and parsed as a new URI.
+//
+// If the URI path is empty or '/', then a duplicate of the URI is returned,
+// along with ErrURIRoot.
+//
+// NOTE: this function should not be called except by an implementation of
+// the Repository interface - using this for unknown URIs may break.
+//
+// Since: 2.0.0
+func GenericParent(u fyne.URI) (fyne.URI, error) {
+	p := u.Path()
+
+	if p == "" || p == "/" {
+		parent, err := ParseURI(u.String())
+		if err != nil {
+			return nil, err
+		}
+		return parent, ErrURIRoot
+	}
+
+	components := splitNonEmpty(u.Path(), "/")
+
+	newURI := u.Scheme() + "://" + u.Authority()
+
+	// there will be at least one component, since we know we don't have
+	// '/' or ''.
+	newURI += "/"
+	if len(components) > 1 {
+		newURI += strings.Join(components[:len(components)-1], "/")
+	}
+
+	// stick the query and fragment back on the end
+	q := u.Query()
+	if len(q) > 0 {
```

Comment:
Any reason this, and below, can't be inlined?

Reply:
f2883d0c66d2de8a94a5bb4236db60653a3fc9eb